Documentation/ceph-encryption.md

```diff
@@ -0,0 +1,225 @@
+  cephVersion:
+    image: quay.io/ceph/ceph:v16.2.6
+    allowUnsupported: false
+  mgr:
+    count: 1
+  storage:
+    storageClassDeviceSets:
+      - name: set1
+        count: 1
+        portable: false
+        tuneDeviceClass: true
+        tuneFastDeviceClass: false
+        encrypted: true    # ==> enable encryption
+```
+
+## Customer requirements & encryption at rest
+There are scenarios when the managers of the Ceph cluster are different than the
+consumers of the cluster. In those scenarios, encryption at rest can be a legal
+and compliance requirement, that includes key rotation support and the ability
+for customers to bring their own encryption keys.
+
+This introduces new feature requirements into rook/ceph:
+1. Support for key rotation.
+2. Support for BYOK for encryption.
+
+## Key Rotation
+As of today, neither Ceph nor Rook support a flag or easy mechanism to rotate a
+Ceph key. Key rotation is feasible through generating a new key and adding it to
+the LUKS header using cryptsetup. There are 2 different approaches to this:
+-  Updating the LUKS key: LUKS header can be updated to change the key, ex:
+   `cryptsetup luksChangeKey --key-file <new key file>` however this doesn't
+   change the `volume key` which is the master key used for encryption.
+-  Updating the master key requires reencryption and can be done through
+   `cryptsetup-reencrypt` which reencrypts data on LUKS device in-place. During
+   reencryption process the LUKS device is marked unavailable.
+   `cryptsetup-reencrypt --key-file <new key file>`.
+
+References:
+- https://man7.org/linux/man-pages/man8/cryptsetup-reencrypt.8.html
+- https://man7.org/linux/man-pages/man8/cryptsetup.8.html#LUKS_EXTENSION
+
+To be able to fulfill key rotation requirements, changes are required to either
+Ceph or Rook depending on the Rook Ceph cluster deployment type:
+
+1.  Host based clusters:
+
+    a.  Ceph: Key rotation will need to be enabled in Ceph and a flag exposed to
+    enabled it. For rotation frequency, it can default to 90 days which is
+    commonly used.
+
+    b.  Rook: Add support to expose and pass the required flags to enable key
+    rotation.
+
+2.  PVC based clusters: Since keys are managed by Rook, it is easier to add such
+    capability into rook. This can be added as an extra flag to the cluster crd
+    `rotateEncryptionKey: true`. Once specified, rook can automatically
+    regenerate a new key on certain frequency (can default to 90days) and update
+    the LUKS header as needed.
+
+
+## BYOK Support in Rook
+
+Customers may be already using existing key management systems to manage their
+secrets or relying on enterprise grade cloud providers for key management. Thus
+the proposal is to leverage  [secrets store
+csi](https://github.com/kubernetes-sigs/secrets-store-csi-driver) for mounting
+of secrets from external secret stores into the OSDs.
+
+Supporting secrets-store-csi will allow customers to pick from the different
```

Comment:
Also, does that mean a single `SecretProviderClass` per OSD? I can see that `SecretProviderClass` has an `[objects](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/release-1.0/test/bats/tests/vault/vault_v1_secretproviderclass.yaml#L10)` list that can have multiple secrets, but it's not like we want all the OSDs secret in it.

Reply:
This is an area that I'm not very sure about. 
It seems from security standpoint we would want to use a different secret per OSD (maintain the current rook implementation), but at the same time this makes secret management a nightmare for customers with large deployments, in this case we may want to use the same key for all OSDs through a single `SecretProviderClass` specified at deployment time. Thoughts?

Reply:
I understand the balance between ease of maintainability and security. The single key per OSD seems a bit more common though. Especially, we lose the key all the OSDs are essentially lost (unless the encrypted disk is still opened). Also, if you decommission one OSD, you can just remove the key and that's it. This makes me a bit nervous to have a single key for all the OSDs...